osf/migrations/0170_populate_file_versions_through.py

```diff
@@ -0,0 +1,98 @@
+        SET
+            DATA TYPE INTEGER;
+        ALTER TABLE osf_basefilenode_versions ALTER COLUMN fileversion_id
+        SET
+            NOT NULL;
+        ALTER TABLE osf_basefilenode_versions ALTER COLUMN fileversion_id
+        SET
+            DATA TYPE INTEGER;
+        ALTER TABLE osf_basefilenode_versions ALTER COLUMN basefilenode_id
+        SET
+            NOT NULL;
+        ALTER TABLE osf_basefilenode_versions ADD CONSTRAINT osf_basefilenode__fileversion_id_93etanfc_fk_osf_fileversion_id FOREIGN KEY (fileversion_id) REFERENCES osf_fileversion DEFERRABLE INITIALLY DEFERRED;
+        ALTER TABLE osf_basefilenode_versions ADD CONSTRAINT osf_basefilenode__fileversion_uniq564 UNIQUE (basefilenode_id, fileversion_id);
+        CREATE INDEX
+        ON osf_basefilenode_versions (basefilenode_id, fileversion_id);
+        CREATE INDEX
+        ON osf_basefilenode_versions (basefilenode_id);
+        CREATE INDEX
+        ON osf_basefilenode_versions (fileversion_id);
+    """
+    with connection.cursor() as cursor:
+        cursor.execute(sql)
+
+
+def populate_fileversion_name(state, schema):
+
+    sql = """
+        DROP TABLE osf_basefileversionsthrough;
+        CREATE TABLE osf_basefileversionsthrough AS
+        SELECT
+            obfv.basefilenode_id,
+            obfv.fileversion_id,
+            ob.name as version_name
+        FROM
+            osf_basefilenode_versions obfv
+            LEFT JOIN
+                osf_basefilenode ob
+                ON obfv.basefilenode_id = ob.id;
+        ALTER TABLE osf_basefileversionsthrough ADD COLUMN id SERIAL PRIMARY KEY;
+        ALTER TABLE osf_basefileversionsthrough ADD CONSTRAINT osf_basefilenod_basefilenode_id_b0nwad27_fk_osf_basefilenode_id FOREIGN KEY (basefilenode_id) REFERENCES osf_basefilenode DEFERRABLE INITIALLY DEFERRED;
+        ALTER TABLE osf_basefileversionsthrough ALTER COLUMN basefilenode_id
+        SET
+            DATA TYPE INTEGER;
+        ALTER TABLE osf_basefileversionsthrough ALTER COLUMN fileversion_id
+        SET
+            NOT NULL;
+        ALTER TABLE osf_basefileversionsthrough ALTER COLUMN fileversion_id
+        SET
+            DATA TYPE INTEGER;
+        ALTER TABLE osf_basefileversionsthrough ALTER COLUMN basefilenode_id
+        SET
+            NOT NULL;
+        ALTER TABLE osf_basefileversionsthrough ADD CONSTRAINT osf_basefilenode__fileversion_id_93nwadfc_fk_osf_fileversion_id FOREIGN KEY (fileversion_id) REFERENCES osf_fileversion DEFERRABLE INITIALLY DEFERRED;
+        ALTER TABLE osf_basefileversionsthrough ADD CONSTRAINT osf_basefilenode__fileversion_uniq UNIQUE (basefilenode_id, fileversion_id);
+        CREATE INDEX
+        ON osf_basefileversionsthrough (basefilenode_id, fileversion_id);
+        CREATE INDEX
+        ON osf_basefileversionsthrough (basefilenode_id);
+        CREATE INDEX
+        ON osf_basefileversionsthrough (fileversion_id);
+    """
+
+    with connection.cursor() as cursor:
+        cursor.execute(sql)
```

Comment:
The gist of all these migrations - there was a default through table linking files and versions and now we need to use a custom one.  1) We add the custom through table model and make a migration.  2) We actually delete that table and use `create table as select` to quickly copy the default table instead + adding the new name fields, and then manually add indices, types, and constraints (to match what we deleted in step 1). 3) Removing the versions field off of the BaseFileNode table (which will delete the default through table) 4) Finally, connecting the BaseFileNodeTable to the custom through table.